freezeyt/freezer.py

```diff
@@ -0,0 +1,150 @@
+import sys
+from pathlib import Path
+from mimetypes import guess_type
+import io
+
+from urllib.parse import urlparse, urljoin
+from werkzeug.datastructures import Headers
+from werkzeug.http import parse_options_header
+
+from freezeyt.encoding import decode_input_path, encode_wsgi_path
+from freezeyt.filesaver import FileSaver
+from freezeyt.dictsaver import DictSaver
+from freezeyt.util import parse_absolute_url, is_external
+from freezeyt.getlinks_html import get_all_links
+from freezeyt.getlinks_css import get_links_from_css
+
+
+def freeze(app, config):
+    freezer = Freezer(app, config)
+    freezer.freeze_extra_files()
+    freezer.handle_urls()
+    return freezer.get_result()
+
+
+def check_mimetype(url_path, headers):
+    if url_path.endswith('/'):
+        # Directories get saved as index.html
+        url_path = 'index.html'
+    f_type, f_encode = guess_type(url_path)
+    if not f_type:
+        f_type = 'application/octet-stream'
+    headers = Headers(headers)
+    cont_type, cont_encode = parse_options_header(headers.get('Content-Type'))
+    if f_type.lower() != cont_type.lower():
+        raise ValueError(
+            f"Content-type '{cont_type}' is different from filetype '{f_type}'"
+            + f" guessed from '{url_path}'"
+        )
+
+
+class Freezer:
+    def __init__(self, app, config):
+        self.app = app
+        self.config = config
+
+        self.extra_pages = config.get('extra_pages', ())
+        self.extra_files = config.get('extra_files', None)
+
+        prefix = config.get('prefix', 'http://localhost:8000/')
+
+        # Decode path in the prefix URL.
+        # Save the parsed version of prefix as self.prefix
+        prefix_parsed = parse_absolute_url(prefix)
+        decoded_path = decode_input_path(prefix_parsed.path)
+        if not decoded_path.endswith('/'):
+            raise ValueError('prefix must end with /')
+        self.prefix = prefix_parsed._replace(path=decoded_path)
+
+        output = config['output']
+        if isinstance(output, str):
+            output = {'type': output}
+
+        if output['type'] == 'dict':
+            self.saver = DictSaver(self.prefix)
+        else:
+            self.saver = FileSaver(Path(output['dir']), self.prefix)
+
+    def get_result(self):
+        get_result = getattr(self.saver, 'get_result', None)
+        if get_result is not None:
+            return self.saver.get_result()
```

Comment:
Ve streamu jsi uváděl, že zde je lepší použít `get_result` jako metodu místo attributu a důvodem jsou  programátorři, kteří tuto metodu budou používat, že jim to naznačí že s tím bude více práce. Mohl by jsi to prosím více rozvést ? Z mého pohledu mi přijde ,že je to úplně jedno a dokonce mi přijde snažší to mít jako attribute

Chápu správně, že pomocí funkce `getattr` uložíme do proměnné definici funkce get_result z objektu `Saver`. Můžeme jinými slovy říct, že zde `getattr` tvoří funkci jakési kontrolní funkce která se dívá jestli existuje definice pro danou funkci a pokud ano tak tento předpis uloží do proměnné `get_result` a pokud ne tak do ní uloží `None` ?
Pokud by to tak opravdu bylo můžeme napsat:

```python
if getattr(self.saver, 'get_result', None) is not None:
    return self.saver.get_result()
```

S tím, že tvá verze je čitelnější a kdybych to chápal správně nechal bych raděj tvůj zápis.

Popřípadě by se mohlo vracet pouze

```python
return get_result()
```